### PR TITLE
perf(gatsby-source-contentful): Add cache for remote imgs

### PR DESCRIPTION
**Pull request by pvdz**
This adds a local cache for the base64 image fetcher.

The tl;dr is that this cuts the build time in about half for at least one site that was fetching an image for the page thumbnail during the "run queries" step.

By default it'll put the images in `.cache/remote_cache/images`.

The assumption is that images are immutable so there's no additional roundtrip in case an image is locally cached.

The local cache is using a sha1 hash for file name because the url length may exceed the file name length on certain (or all) OSes.

Regarding SHA-1: While sha1 is cryptographically insecure, I don't think there's a risk for this purpose. Worst case there is a collision and I'm not quite sure in what scenario you would abuse that.

The PR also adds support for a `GATSBY_REMOTE_CACHE` env var (default is `.cache/remote_cache`), through which you can change the location of the cache folder in case you want to store it away from the Gatsby cache. I find this useful for debugging, where you often `gatsby clean` but still want to keep the remote files.

### CODE REVIEW

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 1]
```diff
@@ -11,10 +15,19 @@ const {
...
 const base64Img = require(`base64-img`)
 const _ = require(`lodash`)
-const path = require(`path`)
 
 const cacheImage = require(`./cache-image`)
 
+// By default store the images in `.cache` but allow the user to override
+// and store the image cache away from the gatsby cache. After all, the gatsby
+// cache is more likely to go stale than the images (which never go stale)
+const REMOTE_CACHE_FOLDER =
```
Comment (KyleAMathews) [inline packages/gatsby-source-contentful/src/extend-node-type.js]: This sort of capability _should_ be built into the core caching APIs. Adding it here could be fine but should be made clear in the comment that this is a one off API & may be removed at any moment.

**Comment by pvdz** [comment]
@KyleAMathews I've added an explicit warning for it, instead

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 2]
```diff
@@ -34,10 +52,26 @@ const getBase64Image = imageProps => {
...
   const requestUrl = `https:${imageProps.baseUrl}?w=20`
-  // TODO add caching.
+
+  // Note: sha1 is unsafe for crypto but okay for this particular case
+  const shasum = crypto.createHash(`sha1`)
+  shasum.update(`requestUrl`)
+  const urlSha = shasum.digest(`hex`)
+
+  // TODO: Find the best place for this step. This is definitely not it.
+  fs.mkdirSync(CACHE_IMG_FOLDER, { recursive: true })
```
Comment (KyleAMathews) [inline packages/gatsby-source-contentful/src/extend-node-type.js]: Any reason not to use the normal cache in the non-remote case? https://www.gatsbyjs.org/docs/build-caching/#the-cache-api

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 3]
```diff
@@ -34,10 +52,26 @@ const getBase64Image = imageProps => {
   if (!imageProps) return null
 
   const requestUrl = `https:${imageProps.baseUrl}?w=20`
-  // TODO add caching.
```
Comment (KyleAMathews) [inline packages/gatsby-source-contentful/src/extend-node-type.js]: heh just noticed this from my initial commit adding this 3 years ago 😆

**Comment by pvdz** [comment]
Going to merge this as-is to unblock certain users. @wardpeet is planning to follow up with a generic solution for this.

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 4]
```diff
@@ -1,3 +1,7 @@
+const fs = require(`fs`)
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/extend-node-type.js]:
```suggestion
const fs = require(`fs-extra`)
```

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 5]
```diff
@@ -11,10 +15,24 @@ const {
...
-const path = require(`path`)
 
 const cacheImage = require(`./cache-image`)
 
+// By default store the images in `.cache` but allow the user to override
+// and store the image cache away from the gatsby cache. After all, the gatsby
+// cache is more likely to go stale than the images (which never go stale)
+// Note that the same image might be requested multiple times in the same run
+
+if (process.env.GATSBY_REMOTE_CACHE) {
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/extend-node-type.js]:
```suggestion
if (process.env.EXPERIMENTAL_GATSBY_REMOTE_CACHE) {
```
